File: Prism.Scripting/src/Prism/Core/Exceptions.cs
using System;

namespace Prism
{
    public class EngineException : Exception
    {
        public EngineException() { }
        public EngineException(string message) : base(message) { }
        public EngineException(string message, Exception inner) : base(message, inner) { }
    }

    public class MissingReferenceException : EngineException
    {
        public MissingReferenceException() : base("The native object has been destroyed or was never created.")
        {
        }

        public MissingReferenceException(string message) : base(message) { }
        public MissingReferenceException(string message, Exception inner) : base(message, inner) { }
    }

    public class EntityNotFoundException : EngineException
    {
        public EntityNotFoundException() : base("The entity does not exist in the active scene.")
        {
        }

        public EntityNotFoundException(ulong entityID)
            : base($"Entity with ID {entityID} not found in the active scene.") { }

        public EntityNotFoundException(string tag)
            : base($"Entity with tag '{tag}' not found in the active scene.") { }

        public EntityNotFoundException(string message, Exception inner) : base(message, inner) { }
    }

    public class ComponentNotFoundException : EngineException
    {
        public ComponentNotFoundException() : base("The component was not found on the entity.")
        {
        }

        public ComponentNotFoundException(Type componentType)
            : base($"Component of type '{componentType.Name}' not found on the entity.") { }

        public ComponentNotFoundException(string message, Exception inner) : base(message, inner) { }
    }

    public class ResourceLoadException : EngineException
    {
        public ResourceLoadException() : base("Failed to load the resource.")
        {
        }

        public ResourceLoadException(string path)
            : base($"Failed to load resource at path: '{path}'") { }

        public ResourceLoadException(string message, Exception inner) : base(message, inner) { }
    }
}

File: PrismEditor/Assets/Scripts/Python/Prism/Core/Exceptions.py
class EngineException(Exception):
    pass


class MissingReferenceException(EngineException):
    def __init__(self, message="The native object has been destroyed or was never created."):
        super().__init__(message)


class EntityNotFoundException(EngineException):
    def __init__(self, message="The entity does not exist in the active scene."):
        super().__init__(message)

    @classmethod
    def by_id(cls, entity_id):
        return cls(f"Entity with ID {entity_id} not found in the active scene.")

    @classmethod
    def by_tag(cls, tag):
        return cls(f"Entity with tag '{tag}' not found in the active scene.")


class ComponentNotFoundException(EngineException):
    def __init__(self, message="The component was not found on the entity."):
        super().__init__(message)

    @classmethod
    def of_type(cls, component_type):
        return cls(f"Component of type '{component_type.__name__}' not found on the entity.")


class ResourceLoadException(EngineException):
    def __init__(self, message="Failed to load the resource."):
        super().__init__(message)

    @classmethod
    def at_path(cls, path):
        return cls(f"Failed to load resource at path: '{path}'")
